File: 820-find-eventual-safe-states/find-eventual-safe-states.cpp
class Solution {
public:

    bool dfs(int i, vector<vector<int>>& graph,unordered_map<int, bool>& safe){
        if (safe.contains(i))
            return safe[i];
        safe[i] = false;
        for (int j: graph[i])
            if (!dfs(j, graph, safe))
                return safe[i];
        safe[i] = true;
        return safe[i];
    }
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> result;
        unordered_map<int, bool> safe;

        for (int i = 0; i < n; i++)
            if (dfs(i, graph, safe))
                result.push_back(i);
        return result;
    }
};

File: 820-find-eventual-safe-states/find-eventual-safe-states.java
class Solution {
    Boolean dfs(int i, int[][] graph, HashMap<Integer, Boolean> safe){
        if (safe.containsKey(i))
            return safe.get(i);
        safe.put(i,false);
        for (int j: graph[i])
            if (!dfs(j, graph, safe))
                return safe.get(i);
        safe.put(i,true);
        return safe.get(i);
    }
    public List<Integer> eventualSafeNodes(int[][] graph) {
        int n = graph.length;
        List<Integer> result = new ArrayList<Integer>();
        HashMap<Integer, Boolean> safe = new HashMap<Integer, Boolean>();

        for (int i = 0; i < n; i++)
            if (dfs(i, graph, safe))
                result.add(i);
        return result;
    }
}